src/dotnet/CarbonAware.LocationSources.Azure/src/AzureLocationSource.cs

```diff
@@ -0,0 +1,107 @@
+using System.Globalization;
+using System.Reflection;
+using CarbonAware.Model;
+using CarbonAware.Interfaces;
+using CarbonAware.Exceptions;
+using Microsoft.Extensions.Logging;
+using System.Text.Json;
+using CarbonAware.LocationSources.Azure;
+
+namespace CarbonAware.LocationSources.Azure;
+
+/// <summary>
+/// Reprsents an azure location source.
+/// </summary>
+public class AzureLocationSource : ILocationSource
+{
+    public string Name => "Azure Location Source";
+
+    public string Description => "Location source that knows how to get and work with Azure location information.";
+
+    private readonly ILogger<AzureLocationSource> _logger;
+
+    private IDictionary<string, NamedGeoposition> namedGeopositions;
+
+    private static readonly JsonSerializerOptions options = new JsonSerializerOptions(JsonSerializerDefaults.Web);
+
+    /// <summary>
+    /// Creates a new instance of the <see cref="AzureLocationSource"/> class.
+    /// </summary>
+    /// <param name="logger">The logger for the datasource</param>
+    public AzureLocationSource(ILogger<AzureLocationSource> logger)
+    {
+        _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        namedGeopositions = new Dictionary<string, NamedGeoposition>();
+    }
+
+    public Location ToGeopositionLocation(Location location)
+    {
+        switch (location.LocationType)
+        {
+            case LocationType.Geoposition:
+            {
+                return location;
+            }
+            case LocationType.CloudProvider: 
+            {
+                if( location.CloudProvider != CloudProvider.Azure ) 
+                {
+                    throw new LocationConversionException($"Incorrect Cloud provider region. Expected Azure but found '{ location.CloudProvider }'");
+                }
+                
+                return getGeoPositionLocationOrThrow(location);
+            }
+        }
+        
+        throw new LocationConversionException($"Location '{ location.CloudProvider }' cannot be converted to Geoposition. ");
+    }
+
+    private Location getGeoPositionLocationOrThrow(Location location)
+    {
+        loadRegionsFromFileIfNotPresent();
+        
+        NamedGeoposition geopositionLocation = namedGeopositions[location.RegionName ?? ""];    
+        if(geopositionLocation == null) 
```

Comment:
Would we also want to invoke this path if the geopositionLocation object is not null but has empty/null members?